lightning/src/util/ser.rs

```diff
@@ -141,6 +296,30 @@ macro_rules! impl_writeable_primitive {
 				Ok($meth_read(&buf))
 			}
 		}
+		impl<R: Read> Readable<R> for HighZeroBytesDroppedVarInt<$val_type> {
+			#[inline]
+			fn read(reader: &mut R) -> Result<HighZeroBytesDroppedVarInt<$val_type>, DecodeError> {
+				// We need to accept short reads (read_len == 0) as "EOF" and handle them as simply
+				// the high bytes being dropped. To do so, we start reading in the middle of buf
+				// and then convert the appropriate number of bytes with extra high bytes out of
+				// buf.
+				let mut buf = [0; $len*2];
+				let mut read_len = reader.read(&mut buf[$len..])?;
+				let mut total_read_len = read_len;
+				while read_len != 0 && total_read_len != $len {
+					read_len = reader.read(&mut buf[($len + read_len)..])?;
+					total_read_len += read_len;
+				}
+				if total_read_len == 0 || buf[$len] != 0 {
```

Comment:
0328440

If I understand correctly, "total_read_len == 0" means here record value is zero, but you should still encode its type instead of skip it ? (but not its length as its null too)

Reply:
This is reading only the V in the TLV - so if we read 0 bytes, it means the L was (probably) 0, so the value is simply 0.

Reply:
But you still have to write down the type if it's a mandatory TLV like amt_to_forward and outgoing_cltv_value right?

Reply:
Hmm? I'm confused what you're asking. This type is only read from a FixedLengthReader that is set to only return the TLV's value field. Further, this can only be used if it is either the last field in a V or the only field in a V, as otherwise there is ambiguity.

Reply:
My question is a onion like `tlv_payload_length | type-2 | type 4` a valid one if you have null `amt_to_forward` and `outgoing_cltv_value`'s L and V ? If yes I don't understand if such HTLC makes sense..

Reply:
Its definitely a valid TLV - this reader (and the TLV tests) test reading a "0 length" such integer, which implies a value of 0.

Reply:
I don't think we sanitize such 0-CLTV/msat  HTLCs  in `send_htlcs` which mean we may send garbage to our outgoing peer and this may trigger undefined behavior ?